src/units/unit.cpp

```diff
@@ -1421,93 +1421,146 @@ void unit::remove_ability_by_id(const std::string& ability)
 
-	if(!bool_or_empty(filter, cfg, "affect_allies"))
-		return false;
+		if(!filter["affect_adjacent"].empty()){
+			bool adjacent = cfg.has_child("affect_adjacent");
+			if(filter["affect_adjacent"].to_bool() != adjacent){
+				return false;
+			}
+		}
 
-	if(!bool_matches_if_present(filter, cfg, "affect_enemies", false))
-		return false;
+		if(!bool_matches_if_present(filter, cfg, "affect_self", true))
+			return false;
 
-	if(!bool_matches_if_present(filter, cfg, "cumulative", false))
-		return false;
+		if(!bool_or_empty(filter, cfg, "affect_allies"))
+			return false;
 
-	const std::vector<std::string> filter_type = utils::split(filter["tag_name"]);
-	if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), tag_name) == filter_type.end() )
-		return false;
+		if(!bool_matches_if_present(filter, cfg, "affect_enemies", false))
+			return false;
 
-	if(!string_matches_if_present(filter, cfg, "id", ""))
-		return false;
+		if(!bool_matches_if_present(filter, cfg, "cumulative", false))
+			return false;
 
-	if(!string_matches_if_present(filter, cfg, "apply_to", "self"))
-		return false;
+		const std::vector<std::string> filter_type = utils::split(filter["tag_name"]);
+		if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), tag_name) == filter_type.end() )
+			return false;
 
-	if(!string_matches_if_present(filter, cfg, "overwrite_specials", "none"))
-		return false;
+		if(!string_matches_if_present(filter, cfg, "id", ""))
+			return false;
 
-	if(!string_matches_if_present(filter, cfg, "active_on", "both"))
-		return false;
+		if(!string_matches_if_present(filter, cfg, "apply_to", "self"))
+			return false;
 
-	//for damage only
-	if(!string_matches_if_present(filter, cfg, "replacement_type", ""))
-		return false;
+		if(!string_matches_if_present(filter, cfg, "overwrite_specials", "none"))
+			return false;
 
-	if(!string_matches_if_present(filter, cfg, "alternative_type", ""))
-		return false;
+		if(!string_matches_if_present(filter, cfg, "active_on", "both"))
+			return false;
 
-	//for plague only
-	if(!string_matches_if_present(filter, cfg, "type", ""))
-		return false;
+		//for damage only
+		if(!string_matches_if_present(filter, cfg, "replacement_type", ""))
+			return false;
 
-	if(!filter["value"].empty()){
-		bool has_other_key = (!cfg["add"].empty() || !cfg["sub"].empty() || !cfg["multiply"].empty() || !cfg["divide"].empty());
-		if(!has_other_key){
-			if(tag_name == "drains"){
-				if(!int_matches_if_present(filter, cfg, "value", 50)){
-					return false;
-				}
-			} else if(tag_name == "berserk"){
-				if(!int_matches_if_present(filter, cfg, "value", 1)){
-					return false;
+		if(!string_matches_if_present(filter, cfg, "alternative_type", ""))
+			return false;
+
+		//for plague only
+		if(!string_matches_if_present(filter, cfg, "type", ""))
+			return false;
+
+		//if ability not return numerical value, and numerical attribute filtered, return false.
```

Comment:
What is this second commit for? Why should the engine know about this, instead of just assuming that the person writing the filter knew what they were filtering for?

Even if this should be here, it needs a comment in the code explaining the purpose, the reason for its existence; a comment that just translates C++ into English isn't saying *why* the code is there.

The list of tags shouldn't be in src/utils, because it's specific to units.

Reply:
second commit removed